crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -437,356 +646,140 @@ where
+        let mut changed_balances =
             self.validate_state_and_get_transfer_data(keys_changed)?;
 
-        let denom = read_denom(&self.ctx.pre(), &transfer.token)?
-            .ok_or_err_msg(
-                "No denomination found in storage for the given token",
-            )?;
+        let masp_address_hash = TransparentAddress(<[u8; 20]>::from(
+            ripemd::Ripemd160::digest(sha2::Sha256::digest(
+                &Address::Internal(Masp).serialize_to_vec(),
+            )),
+        ));
+        verify_sapling_balancing_value(
+            changed_balances
+                .pre
+                .get(&masp_address_hash)
+                .unwrap_or(&ValueSum::zero()),
+            changed_balances
+                .post
+                .get(&masp_address_hash)
+                .unwrap_or(&ValueSum::zero()),
+            &shielded_tx.sapling_value_balance(),
+            epoch,
+            &changed_balances.tokens,
+            conversion_state,
+        )?;
+
+        // The set of addresses that are required to authorize this transaction
+        let mut signers = BTreeSet::new();
+
+        // Checks on the sapling bundle
+        // 1. The spend descriptions' anchors are valid
+        // 2. The convert descriptions's anchors are valid
+        // 3. The nullifiers provided by the transaction have not been
+        // revealed previously (even in the same tx) and no unneeded
+        // nullifier is being revealed by the tx
+        // 4. The transaction must correctly update the note commitment tree
+        // in storage with the new output descriptions
+        self.valid_spend_descriptions_anchor(&shielded_tx)?;
+        self.valid_convert_descriptions_anchor(&shielded_tx)?;
+        self.valid_nullifiers_reveal(keys_changed, &shielded_tx)?;
+        self.valid_note_commitment_update(&shielded_tx)?;
 
-        if transfer.source != Address::Internal(Masp) {
-            // No shielded spends nor shielded conversions are allowed
-            if shielded_tx.sapling_bundle().is_some_and(|bundle| {
-                !(bundle.shielded_spends.is_empty()
-                    && bundle.shielded_converts.is_empty())
-            }) {
-                let error = native_vp::Error::new_const(
-                    "No shielded spends nor shielded conversions are allowed",
-                )
-                .into();
-                tracing::debug!("{error}");
-                return Err(error);
+        // Checks on the transparent bundle, if present
+        validate_transparent_bundle(
+            &shielded_tx,
+            &mut changed_balances,
+            &mut transparent_tx_pool,
+            epoch,
+            conversion_state,
+            &mut signers,
+        )?;
+
+        // Ensure that every account for which balance has gone down has
+        // authorized this transaction
+        for (addr, pre) in changed_balances.pre {
+            if changed_balances.post[&addr] < pre && addr != masp_address_hash {
```

Comment:
Could this check be done only if the balances decrease more than implied by the `Transaction`? As in, if we expect a balance to go down by 10 but we see a decrease of 15 we request a signature since this means that something else happened in parallel to the masp tx that we cannot verify in here. Otherwise, if the decrease matches (or is less than) the exepected value of `Transaction` we don't require/verify the signature. Could this also apply to `validate_transparent_input`? 

. This (if correct) could lead to less signatures needed on the transaction and less verifications.

Reply:
I think you have a point here. I'll try and sketch this change in a separate PR.